File: ga/evaluator.go
package main

type Evaluator interface {
	Evaluate(individual Individual, target []byte) float64
}

type ByteMatch struct{}

func (ByteMatch) Evaluate(individual Individual, target []byte) float64 {
	fitness := 0
	for i, b := range target {
		if b == individual.genome[i] {
			fitness++
		}
	}
	return float64(fitness) / float64(len(target))
}

type CustomEvaluator struct {
	code           string
	needsCompiling bool
	function       func(Individual, []byte) float64
}

func (evaluator *CustomEvaluator) Evaluate(individual Individual, target []byte) float64 {
	if evaluator.function == nil || evaluator.needsCompiling {
		v := codeToFunction(evaluator.code, "Evaluator")
		evaluator.function = v.Interface().(func(individual Individual, target []byte) float64)
		evaluator.needsCompiling = false
	}

	return evaluator.function(individual, target)
}

File: app/src/lib/ts/ga/evaluator.ts
import type { Individual } from "./individual";

export interface Evaluator {
  evaluate(individual: Individual, target: string[]): number;
}

export class CharMatch implements Evaluator {
  evaluate(individual: Individual, target: string[]): number {
    let fitness: number = 0.0;
    target.forEach((char, i) => {
      if (char == individual.genome[i]) fitness++;
    });
    return fitness/target.length;
  }
}

export class CustomEvaluator implements Evaluator {
  evaluate(individual: Individual, target: string[]): number {
    throw new Error("Custom evaluate not implemented");
  }
}